### PR TITLE
[FSDP2] Support input JVP through replicate

### PR DESCRIPTION
**Pull request by weifengpy**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* __->__ #182732

Keep FSDP2 internal unshard and reshard storage bookkeeping outside active functorch transforms while leaving the user module forward under the transform. This also updates the post-backward registration autograd function to use setup_context with identity forward-mode behavior and adds mixed-precision replicate and fully_shard coverage for torch.func.jvp.

Authored by Codex.

Co-authored-by: Birch-san <6141784+Birch-san@users.noreply.github.com>

### CODE REVIEW

**Comment by ezyang** [comment]
Can we discuss what the intended semantics of FSDP2 x JVP are. Because we are running data parallel, to compute the final VJP gradients we have to do an all-reduce (which is what the hook takes care of). With JVP, it is not obvious to me what the intended semantics of the forward gradients are? I guess I'm not sure how the forward grads are getting initialized. Maybe we can check MeanFlow and find out.

**Comment by ezyang** [comment]
Claude on MeanFlow:

```
❯ Talk me through how JVP is used in this repo

⏺ meta_codesearch:code_search(Find JVP usage in repo)
  ⎿  plugin:meta:mux - search_files (MCP)(pattern: "tangent", engine:
                                         "STRMATCH", target_directories:

⏺ meta_codesearch:code_search(Find JVP usage in repo)
  ⎿  Done (31 tool uses · 24.6k tokens · 57s)
  (ctrl+o to expand)

⏺ Here's how JVP is used in this repo:

  Purpose

  JVP (Jacobian-Vector Product) is the core mechanism for computing the total
  time derivative of the model output along a flow trajectory — the key
  mathematical ingredient of the MeanFlow training loss.

  The key computation (meanflow.py:158-170)


  The key computation (meanflow.py:158-170)

  The model predicts a velocity field f(z, t, r, y). During training, JVP computes both the model output u and its
   total time derivative dudt in a single forward pass:

  - Primals: (z, t, r) — noisy input, source timestep, target timestep
  - Tangents: (v_hat, 1, 0) — velocity as the spatial tangent, unit tangent along time, zero for r

  This computes: dudt = du/dz · v_hat + du/dt · 1, i.e. the total derivative of the model output w.r.t. time along
   the flow.

  How it feeds into the loss (meanflow.py:172)

  u_tgt = v_hat - (t - r) * dudt

  This is the MeanFlow consistency condition — the training target is the velocity minus a correction term scaled
  by the time gap.

  Two backend options (meanflow.py:79-103)

  The jvp_api constructor arg selects between:
  - 'funtorch': torch.func.jvp (default, create_graph=False)
  - 'autograd': torch.autograd.functional.jvp (create_graph=True)

  Known limitation (models/dit.py:72)

  Flash/fused attention is explicitly disabled (self.attn.fused_attn = False) because flash attention kernels
  don't support forward-mode AD, which JVP requires. The README notes that NVLabs/rcm has released "JVP in Triton"
   that could lift this restriction.

✻ Crunched for 1m 8s

❯ So what is the differentials I feed in when running the JVP, and does this get fed into a traditional VJP?

⏺ Let me look at the exact code to answer precisely.

  Read 1 file (ctrl+o to expand)

⏺ The tangent vectors

  The tangents at line 164 are:

  (v_hat, torch.ones_like(t), torch.zeros_like(r))

  These correspond to the primals (z, t, r). The physical meaning comes from the flow ODE. Since z = (1-t)*x + t*e
   (line 139), the total derivative of z w.r.t. t is:

  dz/dt = e - x = v

  So the tangent vector represents "what happens as we nudge t by dt":
  - v_hat for z: the spatial direction z moves as t changes (the velocity, possibly CFG-modified)
  - 1 for t: dt/dt = 1
  - 0 for r: r is independent of t, so dr/dt = 0

  Together this computes the total time derivative of the model output along the flow: dudt = (∂u/∂z)·v_hat +
  ∂u/∂t.

  Does it feed into a VJP?

  Yes. Look at lines 172-175:

  u_tgt = v_hat - (t_ - r_) * dudt
  error = u - stopgrad(u_tgt)
  loss = adaptive_l2_loss(error)

  stopgrad blocks gradients through u_tgt, so the backward pass (VJP) only flows through u — the primal output of
  the JVP call, not through dudt. This means:

  - The JVP computes dudt as a forward-mode derivative
  - dudt is used to build the target but is detached (stopgrad)
  - The loss backpropagates through u via standard reverse-mode AD (VJP)

  So the JVP result is not differentiated through. The VJP only sees the normal forward pass output u. This is why
   the 'funtorch' backend can use create_graph=False — there's no need to build a computation graph through the
  JVP tangent computation since dudt never needs gradients flowing back through it.

  The 'autograd' backend with create_graph=True is the more conservative option, but given the stopgrad on u_tgt,
  it shouldn't be necessary in practice.

✻ Churned for 8s

❯ Does this repo implement data parallelism?

⏺ meta_codesearch:code_search(Find data parallelism usage)
  ⎿  Done (15 tool uses · 16.4k tokens · 24s)
  (ctrl+o to expand)

⏺ Yes — via HuggingFace Accelerate, not raw DDP/FSDP.

  Both train.py and train_latent.py follow the same pattern:

  1. Create: accelerator = Accelerator(mixed_precision='fp16')
  2. Wrap: model, optimizer, dataloader = accelerator.prepare(model, optimizer, dataloader) — this handles DDP
  wrapping, device placement, and distributed sampling under the hood
  3. Backward: accelerator.backward(loss) instead of loss.backward()
  4. Guards: accelerator.is_main_process to restrict logging/saving/sampling to rank 0

  The MeanFlow class itself (meanflow.py) is distribution-agnostic — it just computes the loss. All distributed
  concerns are delegated to Accelerate in the training scripts.

✻ Cooked for 30s

❯ In this case, what hapens to the tangent vectors? They would be (implicitly) sharded too?

⏺ Good question. Let me check how the model flows into the loss call.

  Read 1 file (ctrl+o to expand)

⏺ The tangents are not sharded — they're independently computed per-rank from each rank's local data, which is
  exactly the right thing.

  Here's why: DDP replicates the full model on every rank and shards the data batch. So on each rank:

  1. x is the local mini-batch shard (from the distributed sampler)
  2. z = (1-t)*x + t*e, v = e - x, and v_hat are all derived from that local x and local noise e
  3. The tangent vector (v_hat, ones, zeros) is computed locally from local data
  4. The JVP runs entirely locally — forward-mode AD is just a forward pass with dual numbers, no cross-rank
  communication needed
  5. The primal output u and tangent output dudt are both local

  DDP only synchronizes during backward (all-reducing gradients). The JVP is part of the forward pass, so each
  rank independently computes its own u and dudt from its own data shard and its own tangent vectors. The
  gradients from accelerator.backward(loss) get all-reduced across ranks, giving the correct averaged gradient as
  if you'd run on the full batch.

  So there's no special interaction between DDP and JVP here — the tangents are per-sample quantities derived from
   local data, and each rank handles its own independently.

```


And a suggested minimal reproducer:

```
"""Minimal MeanFlow JVP reproducer (~50 LOC)."""
import torch
import torch.nn as nn

# Toy model: predicts velocity field u(z, t, r)
class ToyModel(nn.Module):
    def __init__(self, dim=8):
        super().__init__()
        self.net = nn.Sequential(
            nn.Linear(dim + 2, 64), nn.SiLU(),
            nn.Linear(64, 64), nn.SiLU(),
            nn.Linear(64, dim),
        )

    def forward(self, z, t, r):
        return self.net(torch.cat([z, t, r], dim=-1))

dim = 8
model = ToyModel(dim)
optimizer = torch.optim.Adam(model.parameters(), lr=1e-3)

for step in range(200):
    # Data: random "images" x, noise e
    x = torch.randn(32, dim)
    e = torch.randn(32, dim)

    # Sample timesteps: t (source), r (target), with t > r
    t = torch.rand(32, 1)
    r = t * torch.rand(32, 1)  # r < t

    # Interpolate along flow: z = (1-t)*x + t*e
    z = (1 - t) * x + t * e
    v = e - x  # dz/dt

    # JVP: compute u and its total time derivative dudt simultaneously
    # tangents = (dz/dt, dt/dt, dr/dt) = (v, 1, 0)
    u, dudt = torch.func.jvp(
        lambda z_, t_, r_: model(z_, t_, r_),
        (z, t, r),
        (v, torch.ones_like(t), torch.zeros_like(r)),
    )

    # MeanFlow target: v - (t - r) * dudt
    u_tgt = v - (t - r) * dudt

    # Loss: only backprop through u (primal), not through the target
    loss = ((u - u_tgt.detach()) ** 2).mean()

    optimizer.zero_grad()
    loss.backward()
    optimizer.step()

    if step % 50 == 0:
        print(f"step {step:3d}  loss {loss.item():.4f}")
```

**Review by ezyang — APPROVED** [review]
It would be good to test on the minimal repro, but I think this strategy works.

**Comment by weifengpy** [comment]
> Can we discuss what the intended semantics of FSDP2 x JVP are

Great question! I am treating **input** JVP as orthogonal to FSDP2 parameter sharding. I am not defining semantics for parameter JVP in this PR

The input JVP semantics I have in mind are:
* For a regular single-device `nn.Module`, `torch.func.jvp` computes the directional derivative of the user `module.forward(...)` with respect to the specified inputs
* For `fully_shard(nn.Module)`, FSDP2 adds `unshard -> module.forward(...) -> reshard` around that user forward. Input JVP should see the logical full-parameter user forward, not differentiate through FSDP2-internal unshard/reshard bookkeeping.
* For data-parallel gradient reduction, the parameter gradients are still produced later by `loss.backward()`. Those gradients should go through the normal FSDP2 reduction path

**Comment by weifengpy** [comment]
> Claude on MeanFlow:

thanks for the minimal repro! I was being noob and the repro really helped

I consolidated the minimal MeanFlow-style repro into the unit test. Under `torch.func.jvp`, FSDP2 can see the forward output as `requires_grad=False`. I updated pre-backward hook registration to make the test work

**Comment by Birch-san** [comment]
I believe your minimal repro does not reproduce the problem.

I made a repro:  
https://github.com/pytorch/pytorch/issues/182641#issuecomment-4397294386

I confirm that with your proposed PR: repro.py runs to completion instead of erroring. I don't know about result correctness, but seems a good place to start.

**Comment by weifengpy** [comment]
> 
> I made a repro:  
> 
> https://github.com/pytorch/pytorch/issues/182641#issuecomment-4397294386
> 

thanks. i will incorporate your repro into the unit test

**Comment by weifengpy** [comment]
> I believe your minimal repro does not reproduce the problem.
> 
> I made a repro: [#182641 (comment)](https://github.com/pytorch/pytorch/issues/182641#issuecomment-4397294386)
> 
> I confirm that with your proposed PR: repro.py runs to completion instead of erroring. I don't know about result correctness, but seems a good place to start.

@Birch-san I added you as co-author for the PR. would it be possible for you to sign EasyCLA https://github.com/pytorch/pytorch/pull/182732#issuecomment-4399221661 ?

File: torch/distributed/fsdp/_fully_shard/_fsdp_common.py [hunk 1]
```diff
@@ -28,6 +28,17 @@ def wrapper(*args, **kwargs):
     return wrapper
 
 
+def _disable_functorch_if_active(func):
```
Comment (Skylion007) [inline torch/distributed/fsdp/_fully_shard/_fsdp_common.py]: Please annotated with Callable and ParamSpec / return TypeVar. That way the annotation of the underlying func isn't lost. You can also just do _F with a TypeVar bound to Callable if you prefer.

File: torch/distributed/fsdp/_fully_shard/_fsdp_common.py [hunk 2]
```diff
@@ -28,6 +28,17 @@ def wrapper(*args, **kwargs):
...
+    @functools.wraps(func)
+    def wrapper(*args, **kwargs):
```
Comment (Skylion007) [inline torch/distributed/fsdp/_fully_shard/_fsdp_common.py]: Becomes _P.args and _P.kwargs and return Type R from Callable

File: torch/distributed/fsdp/_fully_shard/_fsdp_common.py [hunk 3]
```diff
@@ -28,6 +28,17 @@ def wrapper(*args, **kwargs):
...
+        if torch._C._are_functorch_transforms_active():
+            with torch._C._DisableFuncTorch():
```
Comment (Skylion007) [inline torch/distributed/fsdp/_fully_shard/_fsdp_common.py:41]: Is this a ContextDecorator? If so you can try decorating the function directly.

File: torch/distributed/fsdp/_fully_shard/_fsdp_param_group.py [hunk 4]
```diff
@@ -1007,14 +1011,26 @@ def _get_param_module_infos(
...
     @staticmethod
     def backward(ctx, *grads: torch.Tensor):
         ctx.param_group.post_backward()
         return (None,) + grads
+
+    @staticmethod
+    def jvp(ctx: Any, *grad_inputs: Any) -> Any:
+        # Drop the non-tensor param_group tangent. The output pre-backward hook
+        # queues final post-backward after all primal/tangent paths finish.
+        return grad_inputs[1:]
```
Comment (Skylion007) [inline torch/distributed/fsdp/_fully_shard/_fsdp_param_group.py]:
```suggestion

from typing import Any
from typing_extensions import TypeVarTuple, Unpack

GradInputs = TypeVarTuple("GradInputs")

@staticmethod
def jvp(
    ctx: Any,
    param_group_tangent: object,
    *grad_inputs: Unpack[GradInputs],
) -> tuple[Unpack[GradInputs]]:
    return grad_inputs
```
You can use this trick to discard the first value.

Comment (weifengpy) [reply torch/distributed/fsdp/_fully_shard/_fsdp_param_group.py]: smart!

File: torch/distributed/fsdp/_fully_shard/_fsdp_common.py (hunk 3 above)
Comment (weifengpy) [reply torch/distributed/fsdp/_fully_shard/_fsdp_common.py:41]: good suggestion. I checked `_DisableFuncTorch` - it cannot be used as `ContextDecorator` since it has no `__call__` implementation

other feedbacks are well covered

**Comment by Birch-san** [comment]
I've signed the CLA, dunno if you need to re-run the EasyCLA check?

**Comment by weifengpy** [comment]
> I've signed the CLA, dunno if you need to re-run the EasyCLA check?

thanks. CLA is working now

**Comment by Skylion007** [comment]
@welfengpy, is it broken through FSDPv1 too?

**Comment by weifengpy** [comment]
> @welfengpy, is it broken through FSDPv1 too?

right, fsdp1 will error on input jvp too. same situation to fsdp2
